zonec.c

```diff
@@ -753,6 +753,426 @@ zparser_conv_nsec(region_type *region,
+		r[3] = htons(port);
+		return r;
+	}
+	zc_error_prev_line("Could not parse port SvcParamValue: \"%s\"", val);
+	return alloc_rdata_init(region, "", 0);
+}
+
+static uint16_t *
+zparser_conv_svcbparam_ipv4hint_value(region_type *region, const char *val)
+{
+	uint16_t *r;
+	int count;
+	char ip_str[INET_ADDRSTRLEN];
+	char *next_ip_str;
+	uint32_t *ip_wire_dst;
+	size_t i;
+
+	for (i = 0, count = 1; val[i]; i++) {
+		if (val[i] == ',')
+			count += 1;
+	}
+
+	/* count == number of comma's in val + 1 
+	 * so actually the number of IPv4 addresses in val
+	 */
+	r = alloc_rdata(region, 2 * sizeof(uint16_t) + IP4ADDRLEN * count);
+	r[1] = htons(SVCB_KEY_IPV4HINT);
+	r[2] = htons(IP4ADDRLEN * count);
+	ip_wire_dst = (void *)&r[3];
+
+	while (count) {
+		if (!(next_ip_str = strchr(val, ','))) {
+			if (inet_pton(AF_INET, val, ip_wire_dst) != 1)
+				break;
+
+			assert(count == 1);
+
+		} else if (next_ip_str - val >= (int)sizeof(ip_str))
+			break;
+
+		else {
+			memcpy(ip_str, val, next_ip_str - val);
+			ip_str[next_ip_str - val] = 0;
+			if (inet_pton(AF_INET, ip_str, ip_wire_dst) != 1) {
+				val = ip_str; /* to use in error reporting below */
+				break;
+			}
+
+			val = next_ip_str + 1;
+		}
+		ip_wire_dst++;
+		count--;
+	}
+	if (count)
+		zc_error_prev_line("Could not parse ipv4hint SvcParamValue: %s", val);
+
+	return r;
+}
+
+static uint16_t *
+zparser_conv_svcbparam_ipv6hint_value(region_type *region, const char *val)
+{
+	uint16_t *r;
+	int i, count;
+	char ip6_str[INET6_ADDRSTRLEN];
+	char *next_ip6_str;
+	uint8_t *ipv6_wire_dst;
+
+	for (i = 0, count = 1; val[i]; i++) {
+		if (val[i] == ',')
+			count += 1;
+	}
+
+	/* count == number of comma's in val + 1 
+	 * so actually the number of IPv6 addresses in val
+	 */
+	r = alloc_rdata(region, 2 * sizeof(uint16_t) + IP6ADDRLEN * count);
+	r[1] = htons(SVCB_KEY_IPV6HINT);
+	r[2] = htons(IP6ADDRLEN * count);
+	ipv6_wire_dst = (void *)&r[3];
+
+	while (count) {
+		if (!(next_ip6_str = strchr(val, ','))) {
+			if ((inet_pton(AF_INET6, val, ipv6_wire_dst) != 1))
+				break;
+
+			assert(count == 1);
+
+		} else if (next_ip6_str - val >= (int)sizeof(ip6_str))
+			break;
+
+		else {
+			memcpy(ip6_str, val, next_ip6_str - val);
+			ip6_str[next_ip6_str - val] = 0;
+			if (inet_pton(AF_INET6, ip6_str, ipv6_wire_dst) != 1) {
+				val = ip6_str; /* for error reporting below */
+				break;
+			}
+
+			val = next_ip6_str + 1; /* skip the comma */
+		}
+		ipv6_wire_dst += IP6ADDRLEN;
+		count--;
+	}
+	if (count)
+		zc_error_prev_line("Could not parse ipv6hint SvcParamValue: %s", val);
+
+	return r;
+}
+
+static int
+network_uint16_cmp(const void *a, const void *b)
+{
+	return ((int)read_uint16(a)) - ((int)read_uint16(b));
+}
+
+static uint16_t *
+zparser_conv_svcbparam_mandatory_value(region_type *region,
+		const char *val, size_t val_len)
+{
+	uint16_t *r;
+	size_t i, count;
+	char* next_key;
+	uint16_t* key_dst;
+
+	for (i = 0, count = 1; val[i]; i++) {
+		if (val[i] == ',')
+			count += 1;
+	}
```

Comment:
It is a good idea to put a max on the count here and fail if too large. Stops too large memory allocations, and 2*count > maxuint16_t.